# Laboratorium 5 — Pandas — Series, DataFrame, eksploracja danych
**Autor:** Katia Kolanko

**Ćwiczenie 1: Series — tworzenie i operacje**

Krok 1 — Series z listy i dict

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Series z listy — domyślny indeks (0, 1, 2, ...)
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


Zadanie 1a: Utwórz Series pensje z dict:

In [4]:
# Klucze: 'Anna', 'Jan', 'Ewa', 'Marek', 'Kasia'
# Wartości: 5500, 7200, 4800, 9100, 6300

pensje = pd.Series({'Anna': 5500, 'Jan': 7200, 'Ewa': 4800, 'Marek': 9100, 'Kasia': 6300})
print(pensje)

Anna     5500
Jan      7200
Ewa      4800
Marek    9100
Kasia    6300
dtype: int64


Zadanie 1b: Odpowiedz na pytania o pensje :

In [5]:
# 1. Wyświetl imię osoby zarabiającej najwięcej (użyj idxmax())
# 2. Oblicz średnią pensję
# 3. Wyświetl imiona osób zarabiających powyżej średniej (filtrowanie boolean)
# 4. Policz ile osób zarabia poniżej 6000 (użyj len() lub .sum())

osoba = pensje.idxmax()
print(osoba)

srednia = pensje.mean()
print(srednia)

maska = pensje > srednia

powyzej_sredniej = pensje[maska]
print(powyzej_sredniej)

ponizej6000 = (pensje < 6000).sum()
print(ponizej6000)


Marek
6580.0
Jan      7200
Marek    9100
dtype: int64
2


Najwięcej zarabia Marek. Średnia pensja wszystkich pracowników wynosi 6580zł. Powyżej średniej zarabia Jan - 7200 i Marek - 9100, natomiast poniżej średniej zarabiają 2 osoby.

Krok 2 — Operacje na Series

Zadanie 1c: Oblicz pensje po zmianach:

In [6]:
# 1. Oblicz pensje po podwyżce 10% dla wszystkich (pensje * 1.10)
# 2. Oblicz pensje po premii 500 zł dla wszystkich (pensje + 500)
# 3. Oblicz nową pensję = (pensja * 1.10) + 500
# 4. Wyświetl ranking nowych pensji od najwyższej (użyj sort_values)

podwyzka10 = pensje * 1.10
print(podwyzka10)

premia = pensje + 500
print(premia)

nowa_pensja = (pensje * 1.10) + 500
print(nowa_pensja)

nowa_pensja.sort_values(ascending=False)

Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0
dtype: float64
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800
dtype: int64
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
dtype: float64


Marek    10510.0
Jan       8420.0
Kasia     7430.0
Anna      6550.0
Ewa       5780.0
dtype: float64

**Ćwiczenie 2: DataFrame — tworzenie i atrybuty**

Krok 1 — DataFrame z dict

In [7]:
# Dane o produktach sklepu internetowego
sklep = pd.DataFrame({
'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
'cena': [3500, 1800, 2500, 350, 2200],
'stan_magazynowy': [45, 120, 200, 500, 75],
'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']})

Zadanie 2a: Zbadaj DataFrame sklep :

In [8]:
# 1. Wyświetl kształt tabeli (shape)
# 2. Wyświetl nazwy kolumn (columns)
# 3. Wyświetl typy danych w każdej kolumnie (dtypes)
# 4. Wyświetl pierwsze 3 wiersze (head)

print(sklep.shape)

print(sklep.columns)

print(sklep.dtypes)

print(sklep.head(3))

(5, 4)
Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria'], dtype='str')
produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object
    produkt  cena  stan_magazynowy  kategoria
0    Laptop  3500               45  Komputery
1    Tablet  1800              120    Mobilne
2  Smartfon  2500              200    Mobilne


Zadanie 2b: Dodaj nowe kolumny do DataFrame sklep :

In [9]:
# 1. Dodaj kolumnę 'wartosc_magazynu' = cena * stan_magazynowy
# 2. Dodaj kolumnę 'cena_z_vat' = cena * 1.23
# 3. Wyświetl cały DataFrame z nowymi kolumnami

sklep['wartosc_magazynu'] = sklep['cena'] * sklep['stan_magazynowy']

sklep['cena_z_vat'] = sklep['cena'] * 1.23

print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


Krok 2 — Wczytanie CSV

In [10]:
# Wczytaj dataset tips
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

Zadanie 2c: Podstawowa eksploracja datasetu tips :

In [11]:
# 1. Wyświetl ile wierszy i kolumn ma dataset (shape)
# 2. Wyświetl ostatnie 5 wierszy (tail)
# 3. Wyświetl 3 losowe wiersze (sample)
# 4. Wyświetl typy danych w każdej kolumnie (dtypes)

print(tips.shape)
print(tips.tail(5))
print(tips.sample(3))
print(tips.dtypes)

(244, 7)
     total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2
     total_bill  tip   sex smoker  day    time  size
28        21.70  4.3  Male     No  Sat  Dinner     2
99        12.46  1.5  Male     No  Fri  Dinner     2
161       12.66  2.5  Male     No  Sun  Dinner     2
total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object


**Ćwiczenie 3: Eksploracja datasetu — samodzielna analiza**

Dane

In [12]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)

Zadania

Zadanie 3a: Pierwszy rzut oka

In [13]:
# 1. Wyświetl head(10)
# 2. Wyświetl shape
# 3. Uruchom info() — ile brakujących wartości widzisz?

print(penguins.head(10))
print(penguins.shape)
penguins.info()

  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       36

Dataset ma 21 brakujących wartości.

Zadanie 3b: Statystyki opisowe

In [ ]:
# 1. Uruchom describe() — co mówią ci średnie i odchylenia?
# 2. Uruchom describe(include='all') — co dodają kolumny tekstowe?
# 3. Jaka jest mediana masy ciała (body_mass_g)?

print(penguins.describe())

print(penguins.describe(include='all'))

print(penguins.body_mass_g)

       bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.459584       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.225000      15.600000         190.000000  3550.000000
50%         44.450000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000
       species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count      344     344      342.000000     342.000000         342.000000   
unique       3       3             NaN            NaN                NaN   
top     Adelie  Biscoe             NaN            NaN                NaN   
freq       152     168             NaN            NaN                NaN   

Średnie wartości pokazują przeciętne cechy pingwinów w zbiorze danych - średnia długość dzioba to około 44mm, średnia masa ciała - 4201g, średnia głębokość dzioba - 17mm i średnia długość płetwy 200mm. Odchylenie standardowe pokazuję jak bardzo wartości różnią się od średniej. Dla długości dzioba jest to 5.45mm, dla głębokości dzioba prawie 2mm, długość płetwy - 14mm i w przypadku masy ciała jest to 801g.

Zadanie 3c: Rozkłady wartości

In [ ]:
# 1. Ile jest pingwinów każdego gatunku? (value_counts na 'species')
# 2. Ile pingwinów na każdej wyspie? (value_counts na 'island')
# 3. Rozkład płci (value_counts na 'sex') — czy jest zbalansowany?
# 4. Jaki procent to gatunek Adelie? (normalize=True)

print(penguins['species'].value_counts())

print(penguins['island'].value_counts())

print(penguins['sex'].value_counts())

print(penguins['species'].value_counts(normalize=True))



species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64
island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64
sex
MALE      168
FEMALE    165
Name: count, dtype: int64
species
Adelie       0.441860
Gentoo       0.360465
Chinstrap    0.197674
Name: proportion, dtype: float64


Najliczniejszym gatunkiem jest Adelie - 152, następnie Gentoo - 124 i Chinstrap - 68. Na wsypie Biscoe znajduję się 168 pingwinów, na wyspie Dream 124, natomiast na wyspie Torgersen jest ich najmniej - 52. Rozkład płci jest zbalansowany liczba samców i samic jest prawie taka sama (168 i 165). Pingwiny Adelie stanowią około 44%.

Zadanie 3d: Brakujące dane

In [19]:
# 1. Policz brakujące wartości w każdej kolumnie (isna().sum())
# 2. Jaki procent danych brakuje w kolumnie 'sex'?
# Podpowiedź: braki / len(penguins) * 100

print(penguins.isna().sum())

braki = penguins['sex'].isna().sum()

procent = braki / len(penguins) * 100

print(procent)




species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64
3.1976744186046515


Najwięcej braków jest w kolumnie sex - 11, a w pozostałych numerycznych kolumnach po 2 braki. W kolumnie 'sex' brakuje około 3% danych.

Zadanie 3e: Wnioski (refleksja)

"Dataset zawiera 344 pingwinów z 3 gatunków..."
"Najczęstszy gatunek to Adelie (152 osobników, 44%)"
"Brakuje danych w kolumnach 'sex', 'bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g' (łącznie 21 braków)"
"Średnia masa ciała to 4201 g (mediana: 4050 g)"
"Pingwiny mieszkają na 3 wyspach, najliczniej na wyspie Biscoe"

**Ćwiczenie 4: Pytania biznesowe + commit**

Dane — wróć do datasetu tips

In [20]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')
# Dodaj kolumnę z procentem napiwku
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)

Pytania

Zadanie 4a:

In [22]:
# 1. Oblicz średni rachunek (total_bill)
# 2. Oblicz średni napiwek w procentach (tip_pct)
# 3. Wyświetl który dzień ma najwięcej rachunków (value_counts na 'day')
# 4. Policz ile rachunków jest od palaczy (value_counts na 'smoker')
# 5. Wyświetl describe() tylko dla kolumny 'tip' — jaka jest mediana napiwku?
# 6. Jaki procent rachunków pochodzi z obiadu (Dinner)?
# Podpowiedź: value_counts(normalize=True) na kolumnie 'time'
# 7. Ile unikalnych rozmiarów grup jest w danych? (nunique na 'size')

print(tips['total_bill'].mean())

print(tips['tip_pct'].mean())

print(tips['day'].value_counts())

print(tips['smoker'].value_counts())

print(tips['tip'].describe()) #mediana - 3

print(tips['time'].value_counts(normalize=True))

print(tips['size'].nunique()) #unikalne rozmiary grup - 6

19.78594262295082
16.08155737704918
day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64
smoker
No     151
Yes     93
Name: count, dtype: int64
count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64
time
Dinner    0.721311
Lunch     0.278689
Name: proportion, dtype: float64
6
